File: src/types/agent.ts
/**
 * Agent type definitions for LangGraph-based agents
 */

export interface Agent {
  id: string;
  organization_id: string;
  name: string;
  description?: string;
  
  // LLM Configuration
  llm_provider: 'openai' | 'anthropic' | 'google' | 'deepseek';
  llm_model: string;
  system_prompt: string;
  temperature: number;
  max_tokens: number;
  
  // Voice Configuration
  voice_provider?: 'elevenlabs' | 'openai' | 'cartesia';
  voice_id?: string;
  voice_model?: string;
  voice_speed: number;
  voice_temperature: number;
  voice_volume: number;
  
  // Speech Processing
  responsiveness: number;
  interruption_sensitivity: number;
  enable_backchannel: boolean;
  backchannel_words: string[];
  normalize_speech: boolean;
  boosted_keywords: string[];
  
  // Ambient & Environment
  ambient_sound: string;
  ambient_volume: number;
  
  // Call Settings
  max_duration_seconds: number;
  end_after_silence_seconds: number;
  voicemail_detection: boolean;
  voicemail_action: 'hangup' | 'leave_message';
  voicemail_message?: string;
  
  // Transcription
  stt_provider: 'whisper' | 'deepgram';
  stt_mode: 'fast' | 'accurate';
  denoising_mode: 'noise-cancellation' | 'noise-and-background-speech-cancellation';
  
  // Security & Privacy
  pii_redaction_enabled: boolean;
  pii_redaction_list: string[];
  data_storage_policy: 'everything' | 'no-pii' | 'basic-attributes';
  
  // Integration
  webhook_url?: string;
  webhook_timeout_ms: number;
  custom_headers: Record<string, string>;
  
  // MCP Tools
  enabled_mcp_servers: string[];
  
  // Knowledge Base
  knowledge_base_ids: string[];
  
  // Status
  is_published: boolean;
  version: number;
  
  // Full config backup
  config_json: AgentConfig;
  
  // Metadata
  created_at: string;
  updated_at: string;
  created_by?: string;
}

export interface AgentConfig {
  name: string;
  description?: string;
  llm_provider: string;
  llm_model: string;
  system_prompt: string;
  temperature: number;
  max_tokens: number;
  voice_provider?: string;
  voice_id?: string;
  voice_model?: string;
  voice_speed: number;
  voice_temperature: number;
  voice_volume: number;
  responsiveness: number;
  interruption_sensitivity: number;
  enable_backchannel: boolean;
  backchannel_words: string[];
  normalize_speech: boolean;
  boosted_keywords: string[];
  ambient_sound: string;
  ambient_volume: number;
  max_duration_seconds: number;
  end_after_silence_seconds: number;
  voicemail_detection: boolean;
  voicemail_action: string;
  voicemail_message?: string;
  stt_provider: string;
  stt_mode: string;
  denoising_mode: string;
  pii_redaction_enabled: boolean;
  pii_redaction_list: string[];
  data_storage_policy: string;
  webhook_url?: string;
  webhook_timeout_ms: number;
  custom_headers: Record<string, string>;
  enabled_mcp_servers: string[];
  knowledge_base_ids: string[];
}

export interface AgentCreateInput {
  name: string;
  description?: string;
  config: AgentConfig;
}

export interface AgentUpdateInput {
  name?: string;
  description?: string;
  config?: Partial<AgentConfig>;
}

export interface AgentSession {
  id: string;
  agent_id: string;
  session_id: string;
  channel: 'text' | 'voice' | 'phone';
  call_id?: string;
  phone_number?: string;
  message_count: number;
  duration_seconds: number;
  status: 'active' | 'ended';
  started_at: string;
  ended_at?: string;
}

export interface AgentMessage {
  id: string;
  session_id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  audio_url?: string;
  audio_duration_ms?: number;
  tokens_used?: number;
  latency_ms?: number;
  created_at: string;
}

File: backend/app/models/agent.py
"""
Agent database model
"""

from sqlalchemy import Column, String, Float, Integer, Boolean, JSON, DateTime, Text
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime
import uuid
from ..database import Base


class Agent(Base):
    """Agent model"""
    __tablename__ = "agents"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    description = Column(Text)
    
    # LLM Configuration
    llm_provider = Column(String(50), nullable=False)
    llm_model = Column(String(100), nullable=False)
    system_prompt = Column(Text, nullable=False)
    temperature = Column(Float, default=0.7)
    max_tokens = Column(Integer, default=1000)
    
    # Voice Configuration
    voice_provider = Column(String(50))
    voice_id = Column(String(255))
    voice_model = Column(String(100))
    voice_speed = Column(Float, default=1.0)
    voice_temperature = Column(Float, default=1.0)
    voice_volume = Column(Float, default=1.0)
    
    # Speech Processing
    responsiveness = Column(Float, default=1.0)
    interruption_sensitivity = Column(Float, default=1.0)
    enable_backchannel = Column(Boolean, default=True)
    backchannel_words = Column(JSON, default=list)
    normalize_speech = Column(Boolean, default=True)
    boosted_keywords = Column(JSON, default=list)
    
    # Ambient & Environment
    ambient_sound = Column(String(50), default='none')
    ambient_volume = Column(Float, default=0.5)
    
    # Call Settings
    max_duration_seconds = Column(Integer, default=1800)
    end_after_silence_seconds = Column(Integer, default=600)
    voicemail_detection = Column(Boolean, default=False)
    voicemail_action = Column(String(20), default='hangup')
    voicemail_message = Column(Text)
    
    # Transcription
    stt_provider = Column(String(50), default='whisper')
    stt_mode = Column(String(20), default='fast')
    denoising_mode = Column(String(50), default='noise-cancellation')
    
    # Security & Privacy
    pii_redaction_enabled = Column(Boolean, default=False)
    pii_redaction_list = Column(JSON, default=list)
    data_storage_policy = Column(String(50), default='everything')
    
    # Integration
    webhook_url = Column(Text)
    webhook_timeout_ms = Column(Integer, default=5000)
    custom_headers = Column(JSON, default=dict)
    
    # MCP Tools
    enabled_mcp_servers = Column(JSON, default=list)
    
    # Knowledge Base
    knowledge_base_ids = Column(JSON, default=list)
    
    # Status
    is_published = Column(Boolean, default=False)
    version = Column(Integer, default=1)
    
    # Full config backup
    config_json = Column(JSON, nullable=False)
    
    # Metadata
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(UUID(as_uuid=True))
